src/lavinmq/amqp/queue/queue.cr

```diff
@@ -190,13 +190,39 @@ module LavinMQ::AMQP
         @paused.set(true)
       end
       @msg_store = init_msg_store(@data_dir)
+      @empty = @msg_store.empty
+      start
+    end
+
+    private def start
       if @msg_store.closed
         close
+      else
+        handle_arguments
+        spawn queue_expire_loop, name: "Queue#queue_expire_loop #{@vhost.name}/#{@name}" if @expires
+        spawn message_expire_loop, name: "Queue#message_expire_loop #{@vhost.name}/#{@name}"
       end
+    end
+
+    def restart!
+      return unless @closed
+      reset_queue_state
+      @msg_store = init_msg_store(@data_dir)
       @empty = @msg_store.empty
-      handle_arguments
-      spawn queue_expire_loop, name: "Queue#queue_expire_loop #{@vhost.name}/#{@name}" if @expires
-      spawn message_expire_loop, name: "Queue#message_expire_loop #{@vhost.name}/#{@name}"
+      start
+    end
+
+    private def reset_queue_state
```

Comment:
I feel that we already have a lot of state to keep track of, while this doesn't add more states it adds more places where we need to keep track of all the states for a queue. 
Not sure if there is a better solution for this at the moment. I just wanted to flag that the more state changes we add to more difficult it will be to keep everything in correct state.